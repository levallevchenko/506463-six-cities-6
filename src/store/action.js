export const ActionType = {
  CHANGE_CITY: `offers/changeCity`,
  CHANGE_SORTING: `offers/changeSorting`,
  LOAD_OFFERS: `offers/loadOffers`,
  LOAD_OFFER_DATA: `offers/loadOfferData`,
  LOAD_REVIEWS: `offers/loadReviews`,
  LOAD_NEARBY_OFFERS: `offers/loadNearbyOffers`,
  SET_COMMENT: `offers/setComment`,
  LOAD_COMMENT: `offers/loadComment`,
  SUBMIT_COMMENT: `offers/submitComment`,
  SET_NOT_FOUND_OFFER: `offers/setNotFoundOffer`,
  SET_OFFERS_LOADING_FAILED: `offers/setOffersLoadingFailed`,
  REQUIRED_AUTHORIZATION: `user/requiredAuthorization`,
  SET_AUTH_INFO: `user/login`,
  LOAD_AUTH_INFO: `user/loadAuthInfo`,
  REDIRECT_TO_ROUTE: `app/redirectToRoute`,
  SET_ERROR: `app/setError`,
};

export const ActionCreator = {
  changeCity: (city) => ({
    type: ActionType.CHANGE_CITY,
    payload: city
  }),
  changeSorting: (sorting) => ({
    type: ActionType.CHANGE_SORTING,
    payload: sorting
  }),
  loadOffers: (offers) => ({
    type: ActionType.LOAD_OFFERS,
    payload: offers,
  }),
  loadOfferData: (offer) => ({
    type: ActionType.LOAD_OFFER_DATA,
    payload: offer,
  }),
  loadReviews: (reviews) => ({
    type: ActionType.LOAD_REVIEWS,
    payload: reviews,
  }),
  loadNearbyOffers: (nearby) => ({
    type: ActionType.LOAD_NEARBY_OFFERS,
    payload: nearby,
  }),
  setComment: (comment) => ({
    type: ActionType.SET_COMMENT,
    payload: comment
  }),
  loadComment: (info) => ({
    type: ActionType.LOAD_COMMENT,
    payload: info
  }),
  submitComment: (info) => ({
    type: ActionType.SUBMIT_COMMENT,
    payload: info
  }),
  setNotFoundOffer: () => ({
    type: ActionType.SET_NOT_FOUND_OFFER,
  }),
  setOffersLoadingFailed: () => ({
    type: ActionType.SET_OFFERS_LOADING_FAILED,
  }),
  setAuthInfo: (info) => ({
    type: ActionType.SET_AUTH_INFO,
    payload: info
  }),
  loadAuthInfo: (info) => ({
    type: ActionType.LOAD_AUTH_INFO,
    payload: info
  }),
  requireAuthorization: (status) => ({
    type: ActionType.REQUIRED_AUTHORIZATION,
    payload: status,
  }),
  redirectToRoute: (url) => ({
    type: ActionType.REDIRECT_TO_ROUTE,
    payload: url
  }),
  setError: (error) => ({
    type: ActionType.SET_ERROR,
    payload: error
  })
};
