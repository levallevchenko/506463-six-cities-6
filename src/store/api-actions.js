import {ActionCreator} from "./action";
import {AuthorizationStatus, APIRoute} from "../const";

export const fetchOffersList = () => (dispatch, _getState, api) => (
  api.get(APIRoute.OFFERS)
    .then(({data}) => dispatch(ActionCreator.loadOffers(data)))
);

export const fetchOffer = (id) => (dispatch, _getState, api) => (
  api.get(`/hotels/${id}`)
    .then(({data}) => dispatch(ActionCreator.loadOfferData(data)))
    .catch(() => dispatch(ActionCreator.setNotFoundOffer()))
);

export const fetchReviews = (id) => (dispatch, _getState, api) => (
  api.get(`/comments/${id}`)
    .then(({data}) => dispatch(ActionCreator.loadReviews(data)))
);

export const fetchNearbyOffers = (id) => (dispatch, _getState, api) => (
  api.get(`/hotels/${id}/nearby`)
    .then(({data}) => dispatch(ActionCreator.loadNearbyOffers(data)))
);

export const postComment = ({review: comment, rating}, id) => (dispatch, _getState, api) => {
  dispatch(ActionCreator.loadComment(true));
  api.post(`/comments/${id}`, {comment, rating})
    .then(({data}) => dispatch(ActionCreator.setComment(data)))
    .catch(() => {});
};

export const checkAuth = () => (dispatch, _getState, api) => (
  api.get(APIRoute.LOGIN)
    .then(({data}) => dispatch(ActionCreator.setAuthInfo(data)))
    .then(() => dispatch(ActionCreator.loadAuthInfo(true)))
    .then(() => dispatch(ActionCreator.requireAuthorization(AuthorizationStatus.AUTH)))
    .catch(() => {})
);

export const login = ({login: email, password}) => (dispatch, _getState, api) => (
  api.post(APIRoute.LOGIN, {email, password})
    .then(({data}) => dispatch(ActionCreator.setAuthInfo(data)))
    .then(() => dispatch(ActionCreator.requireAuthorization(AuthorizationStatus.AUTH)))
    .then(() => dispatch(ActionCreator.redirectToRoute(`/`)))
    .catch(() => {})
);
