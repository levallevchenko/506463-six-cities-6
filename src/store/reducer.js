import {CityName, SortingType, AuthorizationStatus} from '../const';
import {ActionType} from './action';
import {adaptOffersData} from '../services/adapter';

const initialState = {
  activeSorting: SortingType.POPULAR,
  activeCity: CityName.PARIS,
  authorizationStatus: AuthorizationStatus.NO_AUTH,
  authInfo: {},
  isOffersLoaded: false,
  offers: [],
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.CHANGE_CITY:
      return {
        ...state,
        activeCity: action.payload
      };
    case ActionType.CHANGE_SORTING:
      return {
        ...state,
        activeSorting: action.payload
      };
    case ActionType.LOAD_OFFERS:
      return {
        ...state,
        offers: adaptOffersData(action.payload),
        isOffersLoaded: true
      };
    case ActionType.REQUIRED_AUTHORIZATION:
      return {
        ...state,
        authorizationStatus: action.payload
      };
    case ActionType.SET_AUTH_INFO:
      return {
        ...state,
        authInfo: action.payload
      };
    default:
      return state;
  }
};

export {reducer};
