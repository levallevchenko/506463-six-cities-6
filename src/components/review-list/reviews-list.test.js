import React from 'react';
import {render} from '@testing-library/react';
import * as redux from 'react-redux';
import configureStore from 'redux-mock-store';
import {NameSpace} from '../../store/root-reducer';
import ReviewList from './review-list';

const mockStore = configureStore({});
jest.spyOn(redux, `useSelector`);
it(`ReviewList should render correctly`, () => {
  const testReviews = [{id: 1, user: {userId: 1, avatarUrl: `1`}}, {userId: 2, user: {avatarUrl: `2`}}];
  const store = mockStore({
    [NameSpace.APP]: {
      error: false,
    },
    [NameSpace.OFFERS]: {
      reviews: [{id: 1, user: {userId: 1, avatarUrl: `1`}}, {id: 2, user: {userId: 2, avatarUrl: `2`}}],
    },

  });

  const {container} = render(
      <redux.Provider store={store}>
        <ReviewList reviews={testReviews} />
      </redux.Provider>
  );

  expect(container).toMatchSnapshot();
});
