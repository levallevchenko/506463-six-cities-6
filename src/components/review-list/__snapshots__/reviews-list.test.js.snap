// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReviewList should render correctly 1`] = `
<div>
  <h2
    class="reviews__title"
  >
    Reviews Â· 
    <span
      class="reviews__amount"
    >
      2
    </span>
  </h2>
  <ul
    class="reviews__list"
  >
    <li
      class="reviews__item"
    >
      <div
        class="reviews__user user"
      >
        <div
          class="reviews__avatar-wrapper user__avatar-wrapper"
        >
          <img
            alt="Reviews avatar"
            class="reviews__avatar user__avatar"
            height="54"
            src="1"
            width="54"
          />
        </div>
        <span
          class="reviews__user-name"
        />
      </div>
      <div
        class="reviews__info"
      >
        <div
          class="reviews__rating rating"
        >
          <div
            class="reviews__stars rating__stars"
          >
            <span />
            <span
              class="visually-hidden"
            >
              Rating
            </span>
          </div>
        </div>
        <p
          class="reviews__text"
        />
        <time
          class="reviews__time"
          datetime=""
        >
          May 2021
        </time>
      </div>
    </li>
    <li
      class="reviews__item"
    >
      <div
        class="reviews__user user"
      >
        <div
          class="reviews__avatar-wrapper user__avatar-wrapper"
        >
          <img
            alt="Reviews avatar"
            class="reviews__avatar user__avatar"
            height="54"
            src="2"
            width="54"
          />
        </div>
        <span
          class="reviews__user-name"
        />
      </div>
      <div
        class="reviews__info"
      >
        <div
          class="reviews__rating rating"
        >
          <div
            class="reviews__stars rating__stars"
          >
            <span />
            <span
              class="visually-hidden"
            >
              Rating
            </span>
          </div>
        </div>
        <p
          class="reviews__text"
        />
        <time
          class="reviews__time"
          datetime=""
        >
          May 2021
        </time>
      </div>
    </li>
  </ul>
</div>
`;
